zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1849,13 +1924,23 @@ private void checkDataSizeLimit(byte[] data) {
   }
 
   public void watchForData(final String path) {
-    retryUntilConnected(new Callable<Object>() {
-      @Override
-      public Object call() throws Exception {
-        getConnection().exists(path, true);
-        return null;
+    watchForData(path, false);
+  }
+
+  private boolean watchForData(final String path, boolean skipWatchingNodeNotExist) {
+    try {
+      if (skipWatchingNodeNotExist) {
```

Comment:
Will it be cleaner if we put this condition into the retryUntilConnected lamda?

Reply:
It should work too. But I don't see the benefit. They are purely orthogonal. Either way would just work

Reply:
One less "retryUntilConnected" : )

Reply:
Do you sincerely feel this is necessary?

Reply:
I usually don't joke when I reviewing code : )
But if you don't agree it will make the code cleaner, you don't have to change. I don't have a strong preference here.

Reply:
Let us keep it this way then. 

Put this inside the lambda has one drawback as I tried. For the two different code path, the return value is of different type. You can choose either type as we don't use return value. But then the question can drag on as which return value to use? Whey prefer this one over the other one?